### DISCUSSION TITLE
How to pass JavaScript functions from V8 to Rust?

### DISCUSSION THREAD

**Discussion by adhamsalama** [denoland/deno#19231] [Q&A] [answered]
Hello everyone

I am following this tutorial about how to write a custom JavaScript runtime using Deno and Rust
https://deno.com/blog/roll-your-own-javascript-runtime
and I'm stuck on implementing setTimeout, because I don't know how JavaScript functions are passed from V8 to Rust, so I don't know what op_set_timeout function signature would look like.

Can anyone help?

Thanks in advance!

**Comment by bartlomieju** [comment, marked as answer]
Example: https://github.com/denoland/deno/blob/26f42a248f4764f85c1c3c3c511b82a990e4b651/core/ops_builtin_v8.rs#L53-L66.

That said - this is limited to only synchronous ops. You won't be able to pass and call a JS function in an asynchronous op. 

If you want to see how to implement `setTimeout` API watch this talk where I walked through it: https://www.youtube.com/watch?v=xJXQwuTB_7Y

**Reply by adhamsalama** [reply]
Sorry for the late reply, thanks for the answer and the video!

**Reply by adhamsalama** [reply]
@bartlomieju This is great, but can you tell how much more complex implementing setInterval is?
The thing with setInterval is that it returns and intervalId which can be used to cancel that interval using clearInterval.
So, how would that be implemented in a single Rust function?

**Comment by yuhr** [comment]
For future readers like me: `node:inspector` implementation will greatly help.

<https://github.com/denoland/deno/blob/d91658b45e175cf4c934120f76bf707ceeb13aad/ext/node/ops/inspector.rs#L103-L161>

Asynchronous ops are also possible by using `#[op2(async)]` attribute, but `pub async fn` doesn't work for some reason, so use `impl Future<Output = ...>` return type instead. [Rust edition 2024 also requires adding `+ use<>` to the RPIT](<https://doc.rust-lang.org/edition-guide/rust-2024/rpit-lifetime-capture.html>).

**Reply by adhamsalama** [reply]
You re-ignited my spark. I may try to have another go at this. Thanks!
